pkg/netns/netns_linux.go

```diff
@@ -0,0 +1,200 @@
+// the returned handle.
+func New() (NetNSHandle, error) {
+	var innerError error
+	var newNetns NetNSHandle
+	restore, err := saveRestoreCurrentThreadNetns()
+	if err != nil {
+		return newNetns, err
+	}
+	defer restore()
+
+	// this invokes the UNSHARE syscall, which creates a new netns *and moves the current thread*
+	// to that netns. So, after creating and getting a FD, we need to move it BACK (via defer), so that
+	// it looks like we've never left the original netns.
+	fd, err := createAndMoveToNewNetNS()
+	if err != nil {
+		return NetNSHandle{fd: -1}, fmt.Errorf("Unable to create new netns: %w", err)
+	}
+	return NetNSHandle{fd: int(fd)}, innerError
+}
+
+// DoByPinPath is for when a process (CNI plugins, etc) needs to jump into a netns
+// managed/owned by a different process in order to do things, and that other process
+// has "pinned" or bind-mounted the netns to a file.
+//
+// Cilium code should not pin or bind-mount network namespaces it creates, this function is
+// purely for interacting with network namespaces created and managed outside of Cilium.
+func DoByPinPath(netNSPath string, runInside func(*NetNSHandle) error) error {
+	fd, err := unix.Open(netNSPath, unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		return err
+	}
+	// We aren't keeping this and we don't own the pin or the netns it refers to
+	// so defer closure.
+	defer func() {
+		err = errors.Join(err, unix.Close(fd))
+	}()
+
+	localHandle := NetNSHandle{
+		fd: int(fd),
+	}
+	err = localHandle.Do(runInside)
+	return err
+}
+
+// FD returns the underlying file descriptor of the NetNSHandle
+// Note that FDs should not be used for comparing or testing equality, for that
+// see `equals`
+func (h *NetNSHandle) FD() int {
+	return h.fd
+}
+
+// Close closes the handle to the network namespace. Note that this does not guarantee
+// the network namespace itself will be destroyed or closed, which will happen when all references to it
+// (including the one being closed here) are gone.
+func (h *NetNSHandle) Close() error {
+	if err := unix.Close(h.fd); err != nil {
+		return err
+	}
+	h.fd = -1
+	return nil
+}
+
+// Do runs the provided func within a new thread+netns, without changing the current thread's netns
+//
+// Note that since network namespaces are set *per thread*, the thread the closure is running inside
+// will be locked with `runtime.LockOSThread()` during this operation.
+//
+// This means the provided closure should _not_ call runtime.UnlockOSThread() at all,
+// or else undefined behavior should be expected.
+func (h *NetNSHandle) Do(runInside func(*NetNSHandle) error) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// Start the callback in a new green thread so that if we later fail
+	// to switch the namespace back to the original one, we can safely
+	// leave the thread locked to die without a risk of the current thread
+	// left lingering with incorrect namespace.
+	var innerError error
```

Comment:
How about using a `sync.ErrGroup` instead? That avoids the awkward error variable scoping issues.

Reply:
Oh that's new to me. Much better.